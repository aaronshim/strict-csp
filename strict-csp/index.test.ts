// strict-csp/index.test.ts
import { StrictCsp } from './index';
import * as crypto from 'crypto';

describe('StrictCsp.hashInlineScript', () => {
  it('should correctly calculate the sha256 hash of a simple script', () => {
    // Arrange
    const scriptContent = `console.log('hello');`;
    const expectedHash = crypto
      .createHash('sha256')
      .update(scriptContent, 'utf-8')
      .digest('base64');

    // Act
    const result = StrictCsp.hashInlineScript(scriptContent);

    // Assert
    expect(result).toBe(`'sha256-${expectedHash}'`);
  });

  it('should produce a consistent hash for an empty string', () => {
    const result = StrictCsp.hashInlineScript('');
    expect(result).toBe("'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='");
  });

  it('should correctly handle scripts with leading/trailing whitespace', () => {
    // Whitespace is significant in hashes
    const scriptContent = `
      alert('test');
    `;
    const expectedHash = crypto
      .createHash('sha256')
      .update(scriptContent, 'utf-8')
      .digest('base64');

    const result = StrictCsp.hashInlineScript(scriptContent);
    expect(result).toBe(`'sha256-${expectedHash}'`);
  });
});

describe('StrictCsp.getStrictCsp', () => {
    it('should generate a valid strict CSP policy', () => {
        const hashes = [`'sha256-someHash123='`, `'sha256-anotherHash456='`];
        const result = StrictCsp.getStrictCsp(hashes, {
            enableBrowserFallbacks: true,
            enableTrustedTypes: false,
            enableUnsafeEval: false,
        });

        // Using a snapshot for a more complex output
        expect(result).toMatchSnapshot();
    });
});
