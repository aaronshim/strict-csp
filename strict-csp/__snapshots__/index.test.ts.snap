// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StrictCsp.getStrictCsp should generate a valid strict CSP policy 1`] = `"script-src 'strict-dynamic' 'sha256-someHash123=' 'sha256-anotherHash456=' https: 'unsafe-inline';object-src 'none';base-uri 'self';"`;
